### ISSUE TITLE
SystemError building with pyproj

### ISSUE THREAD

**Issue by nchidsey** [py2exe/py2exe#20]
Picking up on discussion from #18, I was able to get a few packages to work by specifying the package names in the options, but one it did not work for is pyproj.

On Python 3.7.x (32 and 64-bit), same setup as in #18, but with `pip install pyproj==2.2.2` to get a windows wheel.  The following is the output of the resulting exe, with a bonus warning about distutils which I don't know is related.  The last import, `pyproj._crs` is a pyc in the zip that loads a pyd outside the zip.

```
...\py2exe_zipimport_bug\dist\library.zip\distutils\__init__.py:15: UserWarning: The virtualenv distutils package at %s appears to be in the same location as the system distutils?
Traceback (most recent call last):
  File "foo.py", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "C:\Users\nick\.virtualenvs\tmp-b1da4736-4c9a-4313-808c-4bdae17f2fdb\lib\site-packages\pyproj\__init__.py", line 77, in <module>
    from pyproj.crs import CRS  # noqa: F401
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "C:\Users\nick\.virtualenvs\tmp-b1da4736-4c9a-4313-808c-4bdae17f2fdb\lib\site-packages\pyproj\crs.py", line 36, in <module>
    from pyproj._crs import CoordinateOperation  # noqa
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
SystemError: <method 'load_module' of 'zipimport.zipimporter' objects> returned NULL without setting an error
```

**Comment by albertosottile** [comment]
Hi, thanks for reporting this. Before proceeding with this issue, I would like to reproduce it. If you could provide me the script and the `setup.py` file that you used to trigger it, that would be great.

**Comment by nchidsey** [comment]
Using Python 3.7.3, 32-bit.  Interesting that this time there is a newer version of pyproj (2.4.0 vs 2.2.2), and the stack is different, but the exception is in the same place.

`setup.py`:
```python
from setuptools import setup
import py2exe
setup(
    name='foo',
    version='1.0',
    console=['foo.py'],
    options={
        'py2exe': {
            'packages': ['pyproj'],
        },
    },
)
```

`foo.py`:
```python
import pyproj
print("ok")
```

venv setup and build:
```
pip install pyproj==2.4.0
pip install https://github.com/albertosottile/py2exe/releases/download/v0.9.3.1/py2exe-0.9.3.1-cp37-none-win32.whl
python setup.py py2exe
```

output of exe:
```
C:\Users\nick\.virtualenvs\tmp-cbec3d49-6cb7-4a7e-a8ec-a184662ed210\dist\library.zip\distutils\__init__.py:15: UserWarning: The virtualenv distutils package at %s appears to be in the same location as the system distutils?
Traceback (most recent call last):
  File "foo.py", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "C:\Users\nick\.virtualenvs\tmp-cbec3d49-6cb7-4a7e-a8ec-a184662ed210\lib\site-packages\pyproj\__init__.py", line 70, in <module>
    from pyproj import _datadir
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
SystemError: <method 'load_module' of 'zipimport.zipimporter' objects> returned NULL without setting an error
```

**Comment by albertosottile** [comment]
Hi, I have recently improved the DLL finder system in `master` and if you could test this change with thisissue, that would be great. Could you install the (not yet released) 0.9.4.0 wheels from [Bintray](https://bintray.com/alby128/py2exe/py2exe/0.9.4.0#files) and tell me if you are still able to reproduce this?  Thanks again for your help.

**Comment by nchidsey** [comment]
Unfortunately, it gave the same exception.
```
Traceback (most recent call last):
  File "foo.py", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
  File "pyproj\__init__.pyc", line 70, in <module>
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 668, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 638, in _load_backward_compatible
SystemError: <method 'load_module' of 'zipimport.zipimporter' objects> returned NULL without setting an error
```

**Comment by albertosottile** [comment]
I am afraid that I am unable to reproduce. This is what I am getting with 0.9.4.0 on cp37-win_amd64:
```
C:\py2exe\tests\functional\pyproj_test\dist>pyproj_test.exe
C:\py2exe\tests\functional\pyproj_test\dist\library.zip\pyproj\__init__.py:90: UserWarning: Valid PROJ data directory not found. Either set the path using the environment
al variable PROJ_LIB or with `pyproj.datadir.set_data_dir`.
ok
```
The proj data directory is not packed at all, and might require a specific hook. Nevertheless, I am not able to reproduce your specific error case.

**Comment by albertosottile** [comment]
By manually copying the `proj` folder and adapting the `pyproj_test.py` script as:
```
import os.path
import pyproj
datadir = os.path.abspath('proj')
pyproj.datadir.set_data_dir(datadir)
pr = pyproj.Proj('esri:102008')
print(pr)
print("ok")
```
it is actually possible to test the software functionality from the packed executable:
```
C:\py2exe\tests\functional\pyproj_test\dist>pyproj_test.exe
C:\py2exe\tests\functional\pyproj_test\dist\library.zip\pyproj\__init__.py:90: UserWarning: Valid PROJ data directory not found. Either set the path using the environment
al variable PROJ_LIB or with `pyproj.datadir.set_data_dir`.
Proj('+proj=aea +lat_0=40 +lon_0=-96 +lat_1=20 +lat_2=60 +x_0=0 +y_0=0 +datum=NAD83 +units=m +no_defs', preserve_units=True)
ok
```
Interestingly, the warning stays because it is generated at import time, before the data_dir is set. This might tell us that the best way to write a hook for `pyproj` is via the `PROJ_LIB` environmental variable. 

Please let me know if you are still able to reproduce the issue reported before. Otherwise, I think I will start working on a hook to copy the data folder and set the env variable at runtime.

**Comment by nchidsey** [comment]
I just updated my Python installation from 3.7.3 to 3.7.4 (32-bit), setup a new virtualenv with it, rebuilt the test, and got the same exception from the exe.  I tried this with both pyproj 2.2.2 and 2.4.0, which gave different tracebacks, but the same SystemError about zipimport.  I'll get a 64-bit python and try with that, but I wonder what the difference is between our setups.

**Comment by albertosottile** [comment]
The difference probably lies in the use of virtualenv. I ran my tests in an ad hoc docker container. Nevertheless, it looks that the problem could be just caused by pyproj not being able to locate the proj folder, so I guess that in any case I should work on a hook for it.

**Comment by nchidsey** [comment]
I'm not sure that the problem for me is the data directory.  The exception is happening when the test script imports `pyproj`, when that imports `pyproj._datadir`.  That's a Cython file, which I think py2exe makes some kind of stub pyc in library.zip to load a pyd that's outside the zip.  I'm not sure exactly what's going wrong in that transition, but I don't *think* it's trying to locate the data directory when it imports, but instead later at the bottom of `__init__.py`.

In any case, I tried setting `PROJ_LIB` both before running setup.py as well as before running the exe, and it did not change the exception.

**Comment by albertosottile** [comment]
Your investigation seems sound, but I still do not know why I cannot reproduce this locally. Let's try this: please set up this environment variable before running the **packed** executable (after running the setup.py script):
`set PY2EXE_VERBOSE=1`.

Then run the executable and report the log here.

**Comment by nchidsey** [comment]
[verbose_log.txt](https://github.com/albertosottile/py2exe/files/3722136/verbose_log.txt)

The distutils warning is still in there. I don't know if that's related.  I wish libraries wouldn't use distutils outside of their setup, causing it to be included in the py2exe build. It tends to cause issues.

**Comment by nchidsey** [comment]
The aforementioned issues occur when distutils tries to import `site`.  I just ran a test making sure `distutils` was in py2exe's `packages` option, and only this in the main test script:
```python
from distutils.spawn import find_executable
```
and that gives an exception about `site` (along with the same warning):
```
C:\Users\nick\Dev\Projects\_tests\py2exe_pyproj_bug\dist\library.zip\distutils\__init__.py:15: UserWarning: The virtualenv distutils package at %s appears to be in the same location as the system distutils?
Traceback (most recent call last):
  File "foo.py", line 1, in <module>
  File "distutils\__init__.pyc", line 37, in <module>
  File "distutils\command\build_ext.pyc", line 20, in <module>
ModuleNotFoundError: No module named 'site'
```

**Comment by nchidsey** [comment]
Sorry for the rapid posts... So the import chain is roughly
* `pyproj`
* `pyproj._datadir` via a .pyc to .pyd something-or-other
* `pyproj.datadir`
* `distutils` (on its way to importing `distutils.spawn`). This file is super different in a virtualenv.
* `distutils.command.build_ext`  see https://github.com/python/cpython/blob/master/Lib/distutils/command/build_ext.py#L20
* `site`

`site` doesn't end up in the py2exe build, which makes sense. I wouldn't want one to.

If I make a mostly empty python module, set a `USER_BASE` variable on it, and assign it to `sys.modules['site']` before importing either distutils or pyproj, then it works around both exceptions.

I see 3 possibly subjective issues:
* pyproj and all other libraries that are importing from distutils at runtime could be requested to not do so.  It's usually for a version number parser or `find_executable` like in this case.
* even when importing modules that aren't present in the final build, there should be a typical import error message rather than whatever NULL thing is happening in zipimport.  That seems like a bug in some C code misusing the Python API.
* possibly that py2exe somehow work when distutils is included while using a virtualenv.  I'd personally rather see the first two issues corrected, but we can't always get third party library authors to agree.

**Comment by albertosottile** [comment]
I really appreciate your truly deep analysis and your proposals, but I am convinced that working around the virtualenv implementation of distutils might be extremely difficult and not so appealing. I am afraid that the lesson to learn from all this is: do not attempt to freeze code from a virtualenv. This is true also for other packaging software, such as PyInstaller or py2app, and I learned this the hard way over the years.

That being said, I agree with one of your proposals, which is:
> even when importing modules that aren't present in the final build, there should be a typical import error message rather than whatever NULL thing is happening in zipimport. That seems like a bug in some C code misusing the Python API.

I am willing to investigate this as it will help debugging a lot of past and future issues. However, I almost know nothing about the `memimporter` implementation currently used by py2exe. I guess I will take this a chance to learn that code, but I believe this effort will take quite some time. Side note: this is not the typical message that you get when a module is missing, so there must be a specific condition that causes this weird message.

In summary, if you agree, I am inclined to stop working on the original topic of this issue, and rather leave it open but as a reminder to find the causes and provide a better error message for this specific exception.

**Comment by nchidsey** [comment]
Sounds okay to me.  I've had success using py2exe and virtualenvs for a long time on Python 2.7, but I'll consider using Docker for it.  I'll also consider opening issues with other libraries hoping they'll be amenable to removing their distutils dependency.  Because even with using a base Python installation (whether or not via Docker), it's partially lucky that it works... at some point the base Python distutils could import from site like the virtualenv one does and then that alternative wouldn't work anymore to run py2exe.

**Comment by albertosottile** [comment]
Alternatively, a creative fix could be adopted to embed the system `distutils` in the frozen package instead of the virtualenv one (see https://github.com/pyinstaller/pyinstaller/pull/4372). I am not sure about how easy this is using the `py2exe` hook architecture. Moreover, I am still convinced that freezing applications from a virtual environment should not be supported for the moment. Viable alternatives, such as Docker or Anaconda exist for solving the same problem without messing with the standard library.

On the other hand, I gave a look at the code involved in your original issue, and it seems that the traceback is all outside of my control. The issue occurs in `zipimport` and is raised in `importlib`. I could try to catch `SystemError` from zipimport, but that seems too broad to point it only to a missing module (and I will never know which one it is). 

In summary, I am inclined to close this issue for now, if you do not have any other suggestions. Nevertheless, I really appreciated your contribution. I am sure that going once through all this will definitely be useful for when a similar issue will be reported again.

**Comment by albertosottile** [comment]
@nchidsey I played a little with the possibility to bypass a virtualenv in `py2exe` and import distutils (and a bunch of other standard library packages) from the main Python installation. Could you download the latest built wheels from https://bintray.com/alby128/py2exe/py2exe/0.9.4.0#files (not the 3.8 one) and repeat the same test using a virtualenv, and let me know what happens? 

Thanks again for your time.

**Comment by nchidsey** [comment]
Looks like that worked.  No distutils warning, and no exception.

**Comment by albertosottile** [comment]
Thanks, I am going to write a hook for handling the `proj` folder as soon as I have some spare time. I will contact you to test it before closing this issue.

**Comment by albertosottile** [comment]
I think this issue was solved in later versions of `virtualenv`, as I now get a vanilla copy of `distutils` in the virtual environment even without using the hook:
```
(venv) C:\>python
Python 3.9.9 (tags/v3.9.9:ccb0e6a, Nov 15 2021, 18:08:50) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import distutils
>>> import numpy
>>> numpy.__file__
'C:\\venv\\lib\\site-packages\\numpy\\__init__.py'
>>> distutils.__file__
'C:\\Python\\lib\\distutils\\__init__.py'
```

Hence, there is no need for having this custom hook and hence I am going to close this issue now. This does not mean that freezing from a virtual environment is recommended but at least it should be possible, now.
